### DISCUSSION TITLE
Building a new controller with stm32f411

### DISCUSSION THREAD

**Discussion by muhammadelmogy** [grblHAL/STM32F4xx#81] [Show and tell]
I am currently in the process of designing and building a controller board around STM32F411 and I would like to make sure my pin assignment for the uC is done without conflicts and can be safely configured in the Firmware.
Here is the pin assignment from Cube IDE

![GRBLHAL_STM32F411](https://user-images.githubusercontent.com/8148383/179421330-22e833e4-308a-4953-a27a-42b75b9b65c0.jpg)

Here is what I am planning to have in the Board(Initial points & can be changed later)

- 5 axes support
- USART output
- RS485 output
- I2C with strobe input
- Spindle speed control (PWM & 0-10V) & spindle enable 
- User Controls

I hope that I can have some feedback so I can proceed with the design.

**Comment by terjeio** [comment]
PWM on PB10 needs code changes in [driver.h](https://github.com/grblHAL/STM32F4xx/blob/0700ea9ddfa2f8686b7aec4cc2835cb49a8a1db6/Inc/driver.h#L172-L211) that should be verified. 

USART1 RX/TX spanning two ports also needs [code changes](https://github.com/grblHAL/STM32F4xx/blob/0700ea9ddfa2f8686b7aec4cc2835cb49a8a1db6/Src/serial.c#L287-L383) and should be verified as well. I do not know if this is allowed by the MCU.

**Comment by muhammadelmogy** [comment]
Regarding PWM ON PB10 I was able to change the code and verified it with an oscilloscope and bCNC, set max speed up to 35000RPM in the configurations and the PWM matched the speed change.
But for the USART I did not yet made any changes in the code to try it out and I do not have the ability right now to test it.
If I am correct,  I see an initialization code for USART3
SERIAL_MOD == 3
Can it be used to Init USART1 and add it under new mod say mod 4 (I really do not know what I am taking about here :D)

**Comment by terjeio** [comment]
> Can it be used to Init USART1 and add it under new mod say mod 4 (I really do not know what I am taking about here :D)

It should be enabled as SERIAL_MOD 1 with alternative pin assignments specified in the board map - [similar to the way it done for I2C](https://github.com/grblHAL/STM32F4xx/blob/0700ea9ddfa2f8686b7aec4cc2835cb49a8a1db6/Src/i2c.c#L33). Use SERIAL_MOD1_ALT_PINMAP as the symbol for selecting it?

**Reply by muhammadelmogy** [reply]
Ok, I am going to give it a try but I do not know how to really test this?
Currently, before I start designing the circuit for the new controller, I am testing pin assignments on one of those blackpills
But I do not know how to test the USART functionally to make sure that the pin assignments work as intended! 

It was easy for other pins but this one is not as direct as the others!

**Reply by muhammadelmogy** [reply]
I made the required changes and added support for the alternative pins using `SERIAL_MOD1_ALT_PINMAP`
Now to use it, I will define in the map.h file the following:
```
#define SERIAL_MOD  1
#define SERIAL_MOD1_ALT_PINMAP
```
I was able to compile and upload the project to the test board without any issues!
The only thing now is a proper way to test, hope you have some sort of (something) that can help me to test it
Can you please [review it](https://github.com/muhammadelmogy/Mlabs-Controller/blob/bc928fd27ef1a8f0441e5ddc3d3d1be11b5e8766/Src/serial.c#L287-L330), and also is there any other places to change?

**Reply by terjeio** [reply]
Since the pins are on different ports you have to call the initializer twice:

```
GPIO_InitTypeDef GPIO_InitStructure = {
        .Mode      = GPIO_MODE_AF_PP,
        .Pull      = GPIO_NOPULL,
        .Speed     = GPIO_SPEED_FREQ_VERY_HIGH,
        .Pin       = GPIO_PIN_15,
        .Alternate = GPIO_AF7_USART1
    };
    HAL_GPIO_Init(GPIOA, &GPIO_InitStructure);

    GPIO_InitStructure .Pin = GPIO_PIN_3;
    HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);
```

**Reply by muhammadelmogy** [reply]
Ok, I fix it [here](https://github.com/muhammadelmogy/Mlabs-Controller/blob/2f858f762f7fd86c2a5fa1833287aca35cfbf09c/Src/serial.c#L292-L308)

Now any idea about how to test the serial port?
Can ioSender for example send some messages using the serial port on the controller?

**Reply by terjeio** [reply]
Comment out [this line](https://github.com/grblHAL/STM32F4xx/blob/bc215a9dca47911567ae96da94e9fad68af3d423/Inc/my_machine.h#L54) to switch comms from USB to UART. You can then test with a terminal program or a sender - via a [USB to UART bridge](https://www.sjoerdlangkemper.nl/2019/03/20/usb-to-serial-uart/).

**Reply by muhammadelmogy** [reply]
It is working, thanks
